### PR TITLE
Fix test to run on all filesystem

### PR DESCRIPTION
**Pull request by vnghia**
@mihaimaruseac 
- Add argument `cloud_path` which is the path for cloud filesystem (nodename for hdfs, bucket for s3/gcs)
- Change `Setup()` to use `CreateDir()` instead of `mkdir()`

### CODE REVIEW

File: tensorflow/c/experimental/filesystem/modular_filesystem_test.cc [hunk 1]
```diff
@@ -85,17 +85,34 @@ class ModularFileSystemTest : public ::testing::TestWithParam<std::string> {
...
-      GTEST_SKIP() << "Cannot create working directory: "
-                   << tensorflow::IOError(root_dir_, error_code);
+    FileSystem* fs = nullptr;
+    Status s = env_->GetFileSystemForFile(root_dir_, &fs);
+    if (fs == nullptr || !s.ok())
+      GTEST_SKIP() << "No filesystem registered: " << s;
+
+    s = fs->CreateDir(root_dir_);
+    if (!s.ok()) {
+      GTEST_SKIP() << "Cannot create working directory: " << s;
```
Comment (mihaimaruseac) [inline tensorflow/c/experimental/filesystem/modular_filesystem_test.cc:115]: We'll have to think how will this affect testing on filesystems which don't support directories at all.

File: tensorflow/c/experimental/filesystem/modular_filesystem_test.cc [hunk 2]
```diff
@@ -85,17 +85,34 @@ class ModularFileSystemTest : public ::testing::TestWithParam<std::string> {
     const std::string test_name = tensorflow::str_util::StringReplace(
         ::testing::UnitTest::GetInstance()->current_test_info()->name(), "/",
         "_", /*replace_all=*/true);
+    // Since we need the tests for cloud filesystem to run on all OSs (Windows,
+    // MacOS, Linux, ...) The path to temp directory must not be dependent on
+    // the OS which runs the tests.
+    const std::string tmp_dir_ =
+        cloud_path_.empty() ? ::testing::TempDir() : "/tmp/";
```
Comment (mihaimaruseac) [inline tensorflow/c/experimental/filesystem/modular_filesystem_test.cc:92]: `::testing::TempDir()` includes `/tmp` on its output. https://github.com/google/googletest/blob/0eea2e9fc63461761dea5f2f517bd6af2ca024fa/googletest/src/gtest.cc#L6401-L6425

We should instead make this a flag which if not set will default to `::testing::TempDir()`.

**Review by mihaimaruseac — COMMENTED** [review]
I'm thinking of a solution to not have `fs->CreateDir` in `SetUp` and will be back. It mostly looks good, thank you

File: tensorflow/c/experimental/filesystem/modular_filesystem_test.cc (hunk 2 above)
Comment (vnvo2409) [reply tensorflow/c/experimental/filesystem/modular_filesystem_test.cc:92]: if this test run on Windows, `::testing::TempDir()` will return `\\tmp\\`, so this test won't work

File: tensorflow/c/experimental/filesystem/modular_filesystem_test.cc (hunk 1 above)
Comment (vnvo2409) [reply tensorflow/c/experimental/filesystem/modular_filesystem_test.cc:115]: We could make this a flag too. However, it seems that there won't be a filesystem which don't support directories at all.

**Review by mihaimaruseac — APPROVED** [review]

**Pull request merged**